### ISSUE TITLE
EvaporateJS Should Support Parallel File Uploads

### ISSUE THREAD

**Issue by bikeath1337** [TTLabs/EvaporateJS#122]
EvaporteJS supports parallel part uploads but does not support parallel file uploads. If 6 files are uploaded at once, and each file is less than the declared part size, then the files are uploaded in sequence, rather than in parallel.

This would be a nice enhancement to the product.

**Comment by SamPotts** [comment]
+1 :+1:

**Comment by tomsaffell** [comment]
Agreed, this would be nice to have, so long as it's only for the case where the files are smaller than the part size (or similar heuristic). Uploading multiple large files in parallel would result in worse outcomes in the case of network outages, client errors, etc.

**Comment by bikeath1337** [comment]
I hear your concern but it's based upon the (incorrect) assumption that max concurrent parts is an attribute of a FileUpload when the max concurrent parts is actually an attribute of the EvaporateJS instance.

In the latter case, one would start uploading 5 parts of the next file if the previous file was only uploading one part.

**Comment by tomsaffell** [comment]
yes, yes, starting on the next file when some 'slots' are not being used would be an enhancement, for both big and small files. I was referring to a different scenario where:

- Two 1 GB files to be uploaded
-  A 1GB/hour uplink
- Computer get shutdown after 1 hour

If we upload the files in series then when the computer is shutdown we have the first file fully uploaded, and the second is just starting.  If we allow them to be uploaded fully in parallel then when the computer is shutdown we have half of each file uploaded, which worse - and it's much much worse if RACR isnt being used.

**Comment by tot-ra** [comment]
+1

**Comment by Masterdanielsan** [comment]
Is there anyone currently working on this? Is this likely to get some attention?

**Comment by bikeath1337** [comment]
I'm working on it now and am working out the kinks.

**Comment by bikeath1337** [comment]
I published this as `r2.0.0-rc.1` for anyone who wants to give it a spin. It includes this feature.

**Comment by cvn** [comment]
I just tried `r2.0.0-rc.3` and parallel uploads worked well for me. I tested maybe 200 small-ish files and didn't have any issues.

I did a little benchmark with my app to see the impact of parallel uploads.

| Version  | Time to upload 20 small (<1k) text files |
| ------------- | ------------- |
| 1.6.1  | 25.6 s  |
| 2.0.0-rc.3  | 10.7 s  |

About 2.4x Faster. This is probably the best-case scenario for parallel uploads, but still pretty cool.

**Comment by bikeath1337** [comment]
Thanks @cvn. I hadn't done any benchmarks yet!

**Comment by tillkruss** [comment]
@cvn: What does your configuration look like for parallel uploads? Did you create two `Evaporate` instances?

**Comment by tillkruss** [comment]
@cvn: Thanks. How exactly can I keep track of the progress of parallel uploads?

I can see the different `progress()` calls, but they don't include any key or file id.

**Comment by bikeath1337** [comment]
@cvn thanks! Good luck @tillkruss
